### ISSUE TITLE
Unable to pass variable objects

### ISSUE THREAD

**Issue by freever** [ChilliCream/hotchocolate#1342]
I have a problem passing objects as variables. 

In this particular case my server exposes users as a Connection that can be filtered, sorted, paginated and queried as a node. The UserConnection is defined as follows:

```c#
descriptor
   .Field("UsersConnection")
   .UsePaging<UserType>()
   .UseFiltering<UserFilterType>()
   .UseSorting<UserSortType>()
   .Resolver((resolverContext, token) =>
   {
      return  resolverContext.Service<UserRepository>()
      .Get(resolverContext.GetAuthenticatedInstruction());
   });
```

The Get() method on UserRepository returns an IQueryable from EF Core that has already been filtered based on some Authentication attributes extracted from the resolverContext. The relevant part of the schema looks as follows:

```graphql
UsersConnection(
after: String
before: String
first: PaginationAmount
last: PaginationAmount
order_by: UserSort
where: UserFilter
): UserConnection
```

The UserFilter object that is passed as the where: parameter consists of the following:

```graphql
type UserFilter {
AND: [UserFilter!]
lastCheckin_gt: DateTime
lastCheckin_lt: DateTime
modifiedAtTicks_gt: Long
OR: [UserFilter!]
totalCheckins_gt: Int
totalCheckins_lt: Int
}
```

For illustration purposes, a super-simple query could just ask for a count of the number of items matching a filter:

```graphql
query GetUserList($filter: UserFilter!) {
  UsersConnection(where: $filter) {
    totalCount
  }
}
```

But when I pass in a variable object to specify the UserFilter:

```graphql
variables:
{
  "filter":  { 
    "totalCheckins_gt": 1
  }
}
```

... I get an error saying "Variable `$filter` got invalid value."

Everything works if I use a scalar property as the input variable instead of an object, i.e:

```graphql
query GetUserList(
	$where_totalCheckins_gt: Int) {
    UsersConnection(
        where: { 
            totalCheckins_gt: $where_totalCheckins_gt
        }) {
    totalCount
  }
}
```

Variables:
```graphql
{
  "where_totalCheckins_gt": 1
}
```

To summarize, it works if I pass in scalar variables and construct the input object inside the query, but if I try and pass in an input object as a variable, it does not work. Am I formatting the input object incorrectly perhaps?

**Comment by freever** [comment]
Some screenshots to illustrate

![works](https://user-images.githubusercontent.com/2269624/72000537-310c1480-324c-11ea-8af7-67fd8d1f3f6b.png)
![works not](https://user-images.githubusercontent.com/2269624/72000538-310c1480-324c-11ea-85fb-79782da39802.png)

**Comment by PascalSenn** [comment]
@freever  I added a test case for this but it passes. Can you see something you did different looking at the test case?

**Comment by freever** [comment]
This is probably a duplicate of #1286 so I'll close that one.

**Comment by freever** [comment]
@PascalSenn I think the problem is how I am deserializing queries before passing them to the QueryExecutor, which is why your (and my) unit tests pass. 

My HC server runs on an Azure function. The HC example of an Azure function does not seem complete so I followed the example here: https://github.com/OneCyrus/GraphQL-AzureFunctions-HotChocolate

Using an Azure function, we need to manually deserialize the incoming query and convert it into a IReadOnlyQueryRequest using QueryRequestBuilder. 

The @OneCyrus sample does this as follows:

```c#
            var graphQLRequest = JsonConvert.DeserializeObject<GraphQLRequest>(await req.ReadAsStringAsync());

            IReadOnlyQueryRequest query = QueryRequestBuilder.New()
                   .SetQuery(graphQLRequest.Query)
                   .SetOperation(graphQLRequest.OperationName).Create();

            var result = await _executor.ExecuteAsync(query);
```

...where GraphQLRequest is:

```c#
    public class GraphQLRequest
    {
        public string OperationName { get; set; }
        public string Query { get; set; }
        public JObject Variables { get; set; }
    }
```

So in the sample no Variables are used as presumbly the use cases so far haven't required them. In order to use Variables I have changed the GraphQLRequest object to use a Dictionary<string, object> instead of JOBject as the Variables property and then I can deserialize the Variables:

```c#
            var query = QueryRequestBuilder.New()
                .SetQuery(graphQlRequest.Instruction.Query)
                .SetOperation(graphQlRequest.Instruction.OperationName)
                .SetVariableValues(graphQlRequest.Instruction.Variables)
                .Create();
```

This works for individual scalar variables, but not for object variables. Objects just deserialize as a json string which then fail to be converted into the expected object input type. 

So that's the problem I am having. So I guess I need to be able to deserialize variables into either a Dictionary<string,object> OR a Dictionary<string,Dictionary<string,object>> but how to do this?

I assume you HC guys have a clever way of doint this outside of the Azure Functions context, could you offer any advice?

**Comment by OneCyrus** [comment]
the processing pipeline in this branch is very basic. we want to create a streamlined AzF helper based on the HC middleware for ASP.Net Core.

I've added a new branch to the repo which uses the HC parser: https://github.com/OneCyrus/GraphQL-AzureFunctions-HotChocolate/tree/feat_hcParser

**Comment by freever** [comment]
resolved by https://github.com/OneCyrus/GraphQL-AzureFunctions-HotChocolate/issues/2
